pkg/network/netlink/consumer.go

```diff
@@ -0,0 +1,404 @@
+	go func() {
+		util.WithRootNS(procRoot, func() {
+			for {
+				fn, ok := <-c.workQueue
+				if !ok {
+					return
+				}
+				fn()
+			}
+		})
+	}()
+}
+
+// do simply dispatches an action to the go-routine running within the root network
+// namespace. the caller can wait for the execution to finish by setting sync to true.
+func (c *Consumer) do(sync bool, fn func()) {
+	if !sync {
+		c.workQueue <- fn
+		return
+	}
+
+	done := make(chan struct{})
+	syncFn := func() {
+		fn()
+		close(done)
+	}
+	c.workQueue <- syncFn
+	<-done
+}
+
+func (c *Consumer) initNetlinkSocket(samplingRate float64) error {
+	var err error
+	c.socket, err = NewSocket()
+	if err != nil {
+		return err
+	}
+
+	c.conn = netlink.NewConn(c.socket, c.socket.pid)
+
+	if err := setSocketBufferSize(netlinkBufferSize, c.conn); err != nil {
+		log.Errorf("error setting rcv buffer size for netlink socket: %s", err)
+	}
+
+	if size, err := getSocketBufferSize(c.conn); err == nil {
+		log.Debugf("rcv buffer size for netlink socket is %d bytes", size)
+	}
+
+	// Attach BPF sampling filter if necessary
+	c.samplingRate = samplingRate
+	atomic.StoreInt64(&c.samplingPct, int64(samplingRate*100.0))
+	if c.samplingRate >= 1.0 {
+		return nil
+	}
+
+	log.Infof("attaching netlink BPF filter with sampling rate: %.2f", c.samplingRate)
+	sampler, _ := GenerateBPFSampler(c.samplingRate)
+	err = c.socket.SetBPF(sampler)
+	if err != nil {
+		atomic.StoreInt64(&c.samplingPct, 0)
+		return fmt.Errorf("failed to attach BPF filter: %w", err)
+	}
+
+	return nil
+}
+
+// receive netlink messages and flushes them to the Event channel.
+// This method gets called in two different contexts:
+//
+// - During system-probe startup, when we're loading all entries from the Conntrack table.
+// In this case c.streaming attribute is false, and once we detect the end of the multi-part
+// message we stop calling socket.Receive() and close the output channel to signal upstream
+// consumers we're done.
+//
+// - When we're streaming new connection events from the netlink socket. In this case, `c.streaming`
+// attribute is true, and only when we detect an EOF we close the output channel.
+// It's also worth noting that in the event of an ENOBUF error, we'll re-create a new netlink socket,
+// and attach a BPF sampler to it, to lower the the read throughput and save CPU.
+func (c *Consumer) receive(output chan Event) {
+ReadLoop:
+	for {
+		buffer := c.pool.Get().(*[]byte)
+		msgs, err := c.socket.ReceiveInto(*buffer)
+
+		if err != nil {
+			switch socketError(err) {
+			case errEOF:
+				// EOFs are usually indicative of normal program termination, so we simply exit
+				return
+			case errENOBUF:
+				// If we detect an ENOBUF during the initial Conntrack table dump it likely means
+				// the netlink socket recv buffer doesn't have enough capacity for the existing connections.
+				if !c.streaming {
+					log.Warnf("netlink: detected enobuf during conntrack table dump. consider raising rcvbuf capacity.")
```

Comment:
I can see this being very bursty, can we omit this since we have a stat for this ?

Reply:
Sure 👍

Reply:
Done.